Straight through notify

gkc opened the pull request:
### NOTE!
This branch includes changes to at_persistence_secondary_server, so at_secondary_server's pubspec.yaml in this branch has a dependency override to get at_persistence_secondary_server from this same branch rather than from pub.dev.

**When this PR is merged**, it should immediately be followed by
- [x] Publish new version of at_persistence_secondary_server (its pupspec.yaml and CHANGELOG.md are ready to go)
- [x] Create a new PR which removes the dependency override for at_persistence_secondary_server from at_secondary_server's pubspec.yaml

### Update, May 12
This is why automated test packs are among the most valuable, if not **_the_** most valuable, of a platform's assets
* I un-skipped the e2e tests for notifications to non-existent destination atSigns, which had previously been marked 'skip' because the tests after them were flaking at the time (for a problem which was subsequently fixed)
* However the un-skipping now caused some flakiness (a different flakiness, caused because I have significantly changed various server config settings - number of retries, quarantine interval, etc)
* Stared at the cicd1 secondary logs on the cicd1 VM and eventually figured out that when connect failed for a given atSign, _processNotificationQueue would not process any other atSign in its current run
* Fixed this so that when connect fails during _processNotificationQueue, then the notifications are re-enqueued for that atSign but the while loop moves on to processing the notifications for the next atSign, if any.

### Main PR description follows
**- What I did**
Removed all avoidable server-side latency under normal conditions when delivering notifications from one atSign to another. (Normal conditions means that both secondary servers are available, connectable and functioning normally)
Lower bound of latency inside the sending secondary server is now approximately 2.5 milliseconds. With regards to total end-to-end latency, for short encrypted messages, the combination of Client-1 encryption time, client-1-to-server-1 communication time, server-2-to-client-2 communication time and client-2 decryption ( excluding other network latency caused by e.g. speed-of-light), adds about another 2.5 milliseconds, giving an overall lower bound for end-to-end latency of approximately 5 milliseconds. 

**- How I did it**
_(All of the 'how I did it' info here comes from this PR's commits' messages. As this PR also contains some logging and de-linting changes, reviewers will likely find it easier to review the changed files one commit at a time)_

In the _at_persistence_secondary_server_ package:
  * in AtNotificationKeystore, publish the new notification event before persisting

In the _at_secondary_server_ package:
* notify_connection_pool.dart
  * added pool.init(size) method. init(size) is called by AtSecondaryImpl when it is starting.
  * pool.get(atSign): Removed redundant second call to _hasCapacity() on the underlying OutboundClientPool
* at_secondary_impl.dart
  * Removed some dead code
  * Moved initialization of the notifications ResourceManager to _after_ the initialization of the OutboundClientManager
  * Uses new init(size) method on the ResourceManager, which in turn calls the new init(size) method on NotifyConnectionPool
* notification_manager_impl.dart
  * NotificationManager_storeNotificationInQueue (called by the public NotificationManager.notify method) now
    1. calls queueManager.enqueue(notification) before persisting it
    2. nudges the notifications ResourceManager to let it know there is a new notification ready to send
* update_verb_handler.dart
  * queues the notification for delivery (via NotificationManager.notify()) before persisting it
  * de-linted
* resource_manager.dart
  * Enhanced the queue processing so that it still periodically runs on a timer, but also can be 'nudged' into action at any time (e.g. NotificationManager and UpdateVerbHandler as described above)
  * Added boolean guard _isProcessingQueue to ensure that the main queue processing loop, in the _processNotificationQueue() function, is not re-entrant
  * Added boolean _nudged so that when _processNotificationQueue() completes, it knows that it should immediately run again
  * Added new init(outboundConnectionsLimit) method (called by AtSecondaryImpl while **it** is starting) which in turns initializes the NotifyConnectionsPool with size _outboundConnectionsLimit
  * Removed this line of code from the main processing for a given atSign within _processNotificationQueue() : `AtNotificationMap.getInstance().removeWaitTimeEntry(atSign);` and moved it into _sendNotifications in order to fix a race-condition bug where
      * if a new notification for a given atSign was enqueued by the NotificationManager at the same time as _processNotificationQueue() was processing previously-queued notifications for **that same atSign** (the window for this to occur is provided by `var outboundClient = await _connect(atSign);` which can be very long)
      * then the next time _processNotificationQueue() would run, with no other notification having been enqueued for that atSign in the interim, it would find nothing to process because the atSign had been removed from the waitTimeMap
  * Changed _sendNotifications so it no longer closes the outbound connection after delivering all of the notifications for a given atSign. Keeping the connection open means that we avoid the (significant) latency added by creating and authenticating a new connection each time.
* at_notification_map.dart : Bug-fix: Made the empty list strongly typed in AtNotificationMap.remove() which was previously untyped and therefore being returned as a ListIterator\<dynamic\>, which was in turn causing problems elsewhere when casting to Iterator\<AtNotification\>
* Made OutboundClientPool usable as an LRU pool
  * Added lastUsed DateTime field to OutboundClient class; field is updated to DateTime.now() every time a method is called
  * Added removeLeastRecentlyUsed() method to OutboundClientPool. Sorts the clients in the pool by lastUsed, ascending
  * Modified NotifyConnectionPool to call removeLeastRecentlyUsed() on its OutboundClientPool if it is at capacity
  * Modified OutboundClientManager to call removeLeastRecentlyUsed() on its OutboundClientPool if it is at capacity
  * Added some tests for direct use of removeLeastRecentlyUsed() method
  * Added some tests on NotifyConnectionsPool to verify LRU behaviour of its underlying OutboundClientPool

**- How to verify it**
All existing and new (in this PR) tests should pass.

**- Description for the changelog**
Removed all avoidable server-side latency under normal conditions when delivering notifications from one atSign to another.

gkc committed:
Changes for faster notifications:
* In AtNotificationKeystore, publish the new notification event before persisting
* Add to AtNotification.toString()

gkc committed:
Changes for faster notifications:
* Updated pubspec.yaml and CHANGELOG.md for at_persistence_secondary_server

gkc committed:
Logging and linting changes

gkc committed:
Changes for faster notifications:
* Updated notification queue processing settings in config.yaml and AtSecondaryConfig.
* Removed some dead code

gkc committed:
Changes for faster notifications:
* Bumped version to 3.0.16, added dependency override for at_persistence_secondary_server to point to at_persistence_secondary_server in this same at_server branch

gkc committed:
Changes for faster notifications:
* queue_manager.dart : de-linted
* handler_util.dart : de-linted
* notification_util.dart : de-linted
* secondary_util.dart : de-linted
* delete_verb_handler.dart : de-linted
* notify_all_verb_handler.dart : de-linted
* notify_verb_handler.dart : de-linted
* update_meta_verb_handler.dart : de-linted

gkc committed:
Changes for faster notifications:
* notify_connection_pool.dart
  * added pool.init(size) method. init(size) is called by AtSecondaryImpl when it is starting.
  * pool.get(atSign): Removed redundant second call to _hasCapacity() on the underlying OutboundClientPool
* at_secondary_impl.dart
  * Removed some dead code
  * Moved initialization of the notifications ResourceManager to _after_ the initialization of the OutboundClientManager
  * Uses new init(size) method on the ResourceManager, which in turn calls the new init(size) method on NotifyConnectionPool

gkc committed:
This commit contains most of the key logic changes.

(The other logic change is in at_persistence_secondary_server in at_notification_keystore.dart, see a few commits ago)

* notification_manager_impl.dart
  * NotificationManager_storeNotificationInQueue (called by the public NotificationManager.notify method) now
    1. calls queueManager.enqueue(notification) before persisting it
    2. nudges the notifications ResourceManager to let it know there is a new notification ready to send
* update_verb_handler.dart
  * queues the notification for delivery (via NotificationManager.notify()) before persisting it
  * de-linted
* resource_manager.dart
  * Enhanced the queue processing so that it still periodically runs on a timer, but also can be 'nudged' into action at any time (e.g. NotificationManager and UpdateVerbHandler as described above)
  * Added boolean guard _isProcessingQueue to ensure that the main queue processing loop, in the _processNotificationQueue() function, is not re-entrant
  * Added boolean _nudged so that when _processNotificationQueue() completes, it knows that it should immediately run again
  * Added new init(outboundConnectionsLimit) method (called by AtSecondaryImpl while **it** is starting) which in turns initializes the NotifyConnectionsPool with size _outboundConnectionsLimit
  * Removed this line of code from the main processing for a given atSign `AtNotificationMap.getInstance().removeWaitTimeEntry(atSign);` as
    1. it is not actually required because of the way getAtSignToNotify(N) works
    2. Removing it fixes a race-condition bug where
      * if a new notification for a given atSign was enqueued by the NotificationManager at the same time as _processNotificationQueue() was processing previously-queued notifications for **that same atSign** (the window for this to occur is provided by `var outboundClient = await _connect(atSign);` which can be very long)
      * then the next time _processNotificationQueue() would run, with no other notification having been enqueued for that atSign in the interim, it would find nothing to process because the atSign had been removed from the waitTimeMap
  * Changed _sendNotifications so it no longer closes the outbound connection after delivering all of the notifications for a given atSign. Keeping the connection open means that we avoid the (significant) latency added by creating and authenticating a new connection each time.
* at_notification_map.dart : Bug-fix: Made the empty list strongly typed in AtNotificationMap.remove() which was previously untyped and therefore being returned as a ListIterator<dynamic>, which was in turn causing problems elsewhere when casting to Iterator<AtNotification>

gkc committed:
removed some logging introduced during dev on this PR

gkc commented:
@cconstab tagging you in your role as bug-finder general to find where this breaks ... branch name is straight_through_notify

gkc committed:
A better fix for the waitTimeMap race condition bug

* During further testing, discovered that while fixing the waitTimeMap race condition bug, I had introduced another (very minor) bug - since the waitTimeMap was not being cleared for a given atSign after successful delivery of all pending notifications for that atSign, then every time _processNotificationQueue would run, it would attempt to connect to the remote secondary even though there were no notifications to send.
* So, instead of just removing `AtNotificationMap.getInstance().removeWaitTimeEntry(atSign);` from _processNotificationQueue(), have instead moved it into _sendNotifications. Have additionally added a check into _sendNotifications to double-check if any new notifications for the atSign have been added to the queue while we were sending the ones that we dequeued, and only clear the waitTimeMap if there are no new notifications.
* Added `length` getters to all_notifications.dart and latest_notifications.dart
* Added `numQueued(atSign)` method to at_notification_map.dart and queue_manager.dart

To recap the original race-condition bug:
      * if a new notification for a given atSign was enqueued by the NotificationManager at the same time as _processNotificationQueue() was processing previously-queued notifications for **that same atSign** (the window for this to occur is provided by `var outboundClient = await _connect(atSign);` which can be very long)
      * then the next time _processNotificationQueue() would run, with no other notification having been enqueued for that atSign in the interim, it would find nothing to process because the atSign had been removed from the waitTimeMap

gkc commented:
Original fix to the waitTimeMap race condition bug introduced a different (albeit very minor) bug. Latest commit fixes that bug also.

gkc commented:
Noticed a potential bug in ResourceManager._enqueueErrorList(). I think that the waitTimeEntry and quarantineEntry for a given atSign should only be cleared when *all* pending notifications have reached maximum retries, but right now it happens if *any* of them have reached maximum retries :
```dart
      if (atNotification.retryCount == maxRetries) {
        AtNotificationMap.getInstance()
            .removeWaitTimeEntry(atNotification.toAtSign);
        AtNotificationMap.getInstance()
            .removeQuarantineEntry(atNotification.toAtSign);
        logger.warning(
            'Failed to notify ${atNotification.id} from ${atNotification.fromAtSign} to ${atNotification.toAtSign}. Maximum retries ($maxRetries) reached');
        continue;
      }
```

Also - for long should we retry with notifications? I've changed the default settings in config.yaml and AtSecondaryConfig so that it's now ~300 seconds (10-second quarantine duration, maxRetries 30) i.e. ~5 minutes rather than ~1 minute as it was previously. But how long should we wait for an unavailable target secondary to become available again?

murali-shris commented on at_secondary/at_secondary_server/lib/src/connection/base_connection.dart:
```diff
@@ -47,11 +47,23 @@ abstract class BaseConnection extends AtConnection {
       throw ConnectionInvalidException('Connection is invalid');
     }
     try {
+      logger.info('SENT: [${getMetaData().sessionID}] ${BaseConnection.truncateForLogging(data)}');
```
@gkc this may cause lot of noise in prod since we log on every socket write. e.g image data will cause lot of encoded chars to be logged

murali-shris commented on at_secondary/at_secondary_server/lib/src/notification/at_notification_map.dart:
```diff
@@ -32,22 +28,32 @@ class AtNotificationMap {
-      return [].iterator as Iterator<AtNotification>;
+  int numQueued(String atSign) {
+    // If map is empty, or map doesn't contain the atSign, return an iterator for an empty list
+    if (_notificationMap.isEmpty || !_notificationMap.containsKey(atSign)) {
```
since we are returning from if block can we remove else block
e.g
if(_notificationMap.isEmpty || !_notificationMap.containsKey(atSign)) {
return 0;
}
// rest of the code

gkc replied:
Sounds good, will do. Makes it a little more readable

murali-shris commented on at_secondary/at_secondary_server/lib/src/notification/resource_manager.dart:
```diff
@@ -11,57 +10,87 @@ import 'package:at_utils/at_logger.dart';
+  void init(int? outboundConnectionLimit) {
+    NotifyConnectionsPool.getInstance().init(outboundConnectionLimit);
+    _isStarted = true;
+    Future.delayed(Duration(milliseconds: 0)).then((value) {
```
if milliseconds is 0..is there a difference in calling _schedule directly vs calling inside Future.delayed(...) ?

gkc replied:
Anecdotally, I believe I've observed a difference but I don't have hard proof. I think it's different in that the code as I've written it creates a new event for the Isolate's event queue

gkc commented:
@murali-shris commented on this pull request.

> In
> at_secondary/at_secondary_server/lib/src/connection/base_connection.dart
> <https://github.com/atsign-foundation/at_server/pull/664#discussion_r864571563>
> :
>
> > @@ -47,11 +47,23 @@ abstract class BaseConnection extends AtConnection {
>        throw ConnectionInvalidException('Connection is invalid');
>      }
>      try {
> +      logger.info('SENT: [${getMetaData().sessionID}] ${BaseConnection.truncateForLogging(data)}');
>
> @gkc <https://github.com/gkc> this may cause lot of noise in prod since
> we log on every socket write. e.g image data will cause lot of encoded
> chars to be logged

Yes, so I added the 'truncateForLogging" method to cut down on the amount of
noise. Having used this configuration for over a week now, I am convinced
that logging all @ protocol interactions (and very little else) at INFO
level makes the logs far more useful.

gkc committed:
More readable AtNotificationMap.add()

gkc committed:
Merge branch 'trunk' into straight_through_notify

gkc committed:
Fixed out-of-order notification delivery

* Made AtNotificationMap.remove() always return a sorted list
* Added a missing `await` in ResourceManager._processNotificationQueue() and changed the return type of _sendNotifications(...) from `void` to `Future<void>` to make it more obvious that it is an async method

gkc committed:
Merge remote-tracking branch 'origin/straight_through_notify' into straight_through_notify

gkc committed:
Fix github workflow

gkc committed:
OutboundClient LRU pool
* Added lastUsed DateTime field to OutboundClient class; field is updated to DateTime.now() every time a method is called
* Added removeLeastRecentlyUsed() method to OutboundClientPool. Sorts the clients in the pool by lastUsed, ascending
* Modified NotifyConnectionPool to call removeLeastRecentlyUsed() on its OutboundClientPool if it is at capacity
* Modified OutboundClientManager to call removeLeastRecentlyUsed() on its OutboundClientPool if it is at capacity
* Some changes to method names and argument names for readability
* Made some variables non-nullable as they are, in fact, not nullable
* Some changes to remove some lint warnings
* Added some tests for removeLeastRecentlyUsed() method
* Sped up some other unit tests (by setting shorter ttb and ttl values)

gkc committed:
More unit tests for LRU outbound pool
* Added some tests on NotifyConnectionsPool to verify LRU behaviour of its underlying OutboundClientPool

gkc commented:
Hi @murali-shris Before I merge to trunk, please can you do another review please, of the commits since May 5th?

gkc committed:
Merge branch 'trunk' into straight_through_notify

gkc committed:
Adjusting e2e tests' config.yaml for server-side notifications handling

gkc committed:
Removed 'skip' from two tests
Removed skips from two tests which had been marked skip in order not to trigger a particular issue with outbound notifications which has since been fixed.

gkc committed:
It's the wrong yaml, Gromit

gkc committed:
And this is why we have e2e tests.
Un-skipping the e2e tests for notifications to non-existent destination atSigns caused some flakiness. Stared at the cicd1 secondary logs on the cicd1 VM and eventually figured out that when connect failed for a given atSign, _processNotificationQueue would not process any other atSign in its current run. Fixed this so that when connect fails during _processNotificationQueue, then the notifications are re-enqueued for that atSign but the while loop moves on to processing the notifications for the next atSign, if any.

gkc committed:
tweaked a delay in a test

gkc committed:
Merge branch 'trunk' into straight_through_notify

gkc committed:
Tweaking config for a specific edge test case

gkc committed:
Tweaking config for a specific edge test case

gkc committed:
Merge remote-tracking branch 'origin/straight_through_notify' into straight_through_notify

gkc committed:
Tweaking config for a specific edge test case

gkc committed:
Trigger a check run
